### PR TITLE
Apply include/exclude configuration to Jacoco Report task

### LINKED ISSUE #14760: [Jacoco Plugin] jacocoTestReport task not honoring class exclusions
We can exclude some classes from code coverage analysis using the following config(I'm excluding 'resource' folder):

**build.gradle:**
```groovy
test {
        useJUnitPlatform()
        finalizedBy jacocoTestReport
        jacoco {
            excludes = ['**/resource/**']
        }
    }
jacocoTestReport {
        dependsOn test
        reports {
            html.enabled true
            xml.enabled true
        }
    }
```
We can them generate the jacoco HTML report with `gradle jacocoTestReport`

### Expected Behavior
- The jacoco should not analyze the classes inside **resource** folder. 
- The HTML report should not include the classes that were excluded from the analysis.

### Current Behavior
- The jacoco works fine and does NOT analyze the classes inside **resource** folder. 
- The HTML report is generated, but it includes the classes that were excluded from the analysis. Those classes show 0% coverage since the analysis task excluded them.

### Context
This is an issue cause the metrics in the coverage report are being incorrectly displayed (The 0% coverage from the excluded classes affects the total coverage metric)

### Steps to Reproduce 
- For any project, configure test and jacocoTestReport tasks as above;
- Run jacocoTestReport task
- Check the generate HTML report

### Your Environment
**Gradle**
```
------------------------------------------------------------
Gradle 6.5
------------------------------------------------------------

Build time:   2020-06-02 20:46:21 UTC
Revision:     a27f41e4ae5e8a41ab9b19f8dd6d86d7b384dad4

Kotlin:       1.3.72
Groovy:       2.5.11
Ant:          Apache Ant(TM) version 1.10.7 compiled on September 1 2019
JVM:          1.8.0_201 (Oracle Corporation 25.201-b09)
OS:           Linux 4.15.0-118-generic amd64
```
**Jacoco**: 0.8.5

### PR DESCRIPTION
**Pull request by fcappi**
Issue: #14760
Signed-off-by: Fernando Cappi <3818944+fcappi@users.noreply.github.com>

Fixes #14760

### Context
See details on the issue #14760

### Contributor Checklist
- [x] [Review Contribution Guidelines](https://github.com/gradle/gradle/blob/master/CONTRIBUTING.md)
- [x] Make sure that all commits are [signed off](https://git-scm.com/docs/git-commit#Documentation/git-commit.txt---signoff) to indicate that you agree to the terms of [Developer Certificate of Origin](https://developercertificate.org/).
- [x] Make sure all contributed code can be distributed under the terms of the [Apache License 2.0](https://github.com/gradle/gradle/blob/master/LICENSE), e.g. the code was written by yourself or the original code is licensed under [a license compatible to Apache License 2.0](https://apache.org/legal/resolved.html).
- [x] Check ["Allow edit from maintainers" option](https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/) in pull request so that additional changes can be pushed by Gradle team
- [ ] Provide integration tests (under `<subproject>/src/integTest`) to verify changes from a user perspective
- [ ] Provide unit tests (under `<subproject>/src/test`) to verify logic
- [ ] Update User Guide, DSL Reference, and Javadoc for public-facing changes
- [x] Ensure that tests pass locally: `./gradlew <changed-subproject>:check`

### Gradle Core Team Checklist
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation
- [ ] Recognize contributor in release notes

### CODE REVIEW

**Comment by JLLeitschuh** [comment]
@fcappi Please add integration tests covering this change

File: subprojects/jacoco/src/main/java/org/gradle/testing/jacoco/tasks/JacocoReportBase.java [hunk 1]
```diff
@@ -181,7 +182,11 @@ public void executionData(TaskCollection tasks) {
      */
     @Internal
     public FileCollection getAllClassDirs() {
-        return classDirectories.plus(getAdditionalClassDirs());
+        FileCollection classes = classDirectories.plus(getAdditionalClassDirs());
+        JacocoTaskExtension jacocoTaskExtension = getProject().getTasks().named(JavaPlugin.TEST_TASK_NAME).get()
+            .getExtensions().getByType(JacocoTaskExtension.class);
+        return classes.getAsFileTree().matching(patterns ->
+            patterns.exclude(jacocoTaskExtension.getExcludes()).include(jacocoTaskExtension.getIncludes()));
```
Comment (JLLeitschuh) [inline subprojects/jacoco/src/main/java/org/gradle/testing/jacoco/tasks/JacocoReportBase.java:189]: I think this needs to be configured at the plugin level, not here.

Comment (JLLeitschuh) [reply subprojects/jacoco/src/main/java/org/gradle/testing/jacoco/tasks/JacocoReportBase.java:189]: Your goal is to configure what the convention is. But users should be able to override the convention. They way you've done it here, you lock in the convention.

Comment (JLLeitschuh) [reply subprojects/jacoco/src/main/java/org/gradle/testing/jacoco/tasks/JacocoReportBase.java:189]: I think the place you want to add this filter is to somewhere in `JacocoPlugin::addDefaultReportTask`.

**Comment by JLLeitschuh** [comment]
@fcappi Any changes here should be driven via an example provided in the integration tests as an example of what's currently broken and what this change will fix.

**Comment by ljacomet** [comment]
The Jacoco plugin needs a deeper look and refactoring. The team is not convinced we can fix such issues until that work has happened.

**Pull request closed without merging**